tensorboard/components/tensor_widget/types.ts

```diff
@@ -0,0 +1,196 @@
+ * should have the following concrete value:
+ *
+ * ```
+ * {
+ *   slicingDimsAndIndices: [{
+ *     dim: 0,
+ *     index: 0
+ *   }, {
+ *     dim: 3,
+ *     index: 2
+ *   }],
+ *   viewingDims: [1, 2],
+ *   verticalRange: [0, 4],
+ *   horizontalRange: [0, 3]
+ * }
+ * ```
+ */
+export interface TensorViewNavigationStatus {
+  /**
+   * Which dimensions of the tensor are sliced down to a slice of 1.
+   *
+   * - The `dim` field is the 0-based dimension index.
+   * - The `index` is the 0-based index for the selected slice.
+   */
+  slicingDimsAndIndices: Array<{dim: number, index: number}>;
+
+  /**
+   * Which dimensoins are used for viewing (i.e., rendered in the
+   * tensor widget, as a table, by default.)
+   */
+  viewingDims: number[];
+
+  /**
+   * The indices from the first viewing dimension, which are shown as rows.
+   *
+   * The two numbers are beginning index (inclusive) and ending index
+   * (exclusive).
+   */
+  verticalRange: [number, number];
+
+  /**
+   * The indices from the second viewing dimension, which are shown as columns.
+   *
+   * The two numbers are beginning index (inclusive) and ending index
+   * (exclusive).
+   */
+  horizontalRange: [number, number];
+
+  /**
+   * Optional dimension for depth.
+   *
+   * This supports visualization that requires a depth dimension, e.g.,
+   * color channels.
+   */
+  depthDim?: number;
+}
+
+/** Options used during the creation of a single-tensor tensor widget. */
+export interface SingleTensorWidgetOptions {
+  /**
+   * Name of the tenosr.
+   *
+   * Optional. If provided, it will be shown on the widget.
+   * If not provided, no name will be shown.
+   */
+  name?: string;
+
+  /**
+   * Whether the health-pill portion of the tensor widget is to be
+   * included
+   *
+   * Defaults to `true`.
+   */
+  includeHealthPill?: boolean;
+
+  /** Defaults to `true`. */
+  includeMenu?: boolean;
+
+  /**
+   * How many decimal places to display the values in.
+   *
+   * The values of the tensor may be displaced in the decimal notation, or
+   * the engineering notation, depending automatically by the tensor-widget
+   * library based on the maximum absolute value of the elements of the tensor.
+   */
+  decimalPlaces?: number;
+
+  /** TODO(cais): Add support for custom tensor renderers. */
+}
+
+/**
+ * A TensorWidget that interactively visuzlies a single tensor.
+ */
+export interface SingleTensorWidget {
+  /**
+   * Renders the GUI of the tensor widget.
+   *
+   * This method should be called only once after the tensor widget is
+   * instantiated, or when the content of the underlying tensor has
+   * changed.
+   */
+  render: () => Promise<void>;
+
+  /** Clears the GUI. */
+  clear: () => Promise<void>;
+
+  /**
+   * Scroll along the horizontal dimension.
+   *
+   * I.e., whichever dimension that's selected as the horizontal viewing
+   * dimension at the current time.
+   */
+  scrollHorizontally: (offset: number) => Promise<void>;
```

Comment:
`offset` will become the first element in the view, regardless of whether the element is already in the view. Added the info to the doc.

Reply:
Strictly speaking there is presumably the edge case in which `offset` specifies the last element and more than one element fits in the widget viewport, in which case it won't actually be the first element (similar to jumping to a header at the bottom of a page where it can't actually scroll all the way down).  Not sure if there's an easy way to integrate that edge case into the docstring.

Reply:
Yes, agreed. I was thinking about how to convey that in the doc string in a succinct way. Since this is somewhat deep in the implementation details, I'd like to leave it to later.